File: 2024/day01/c/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINES 1000
#define HASH_SIZE 100003  // Prime number for better distribution

// Hash table entry
typedef struct {
    int key;
    int count;
    int used;
} HashEntry;

// Hash function
unsigned int hash(int key) {
    return (unsigned int)key % HASH_SIZE;
}

// Insert or increment count in hash table
void hash_insert(HashEntry *table, int key) {
    unsigned int idx = hash(key);

    // Linear probing for collision resolution
    while (table[idx].used && table[idx].key != key) {
        idx = (idx + 1) % HASH_SIZE;
    }

    if (!table[idx].used) {
        table[idx].key = key;
        table[idx].count = 1;
        table[idx].used = 1;
    } else {
        table[idx].count++;
    }
}

// Get count from hash table
int hash_get(HashEntry *table, int key) {
    unsigned int idx = hash(key);

    // Linear probing to find the key
    while (table[idx].used) {
        if (table[idx].key == key) {
            return table[idx].count;
        }
        idx = (idx + 1) % HASH_SIZE;
    }

    return 0;  // Not found
}

// Comparison function for qsort
int compare_ints(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

int main() {
    FILE *file = fopen("../input.txt", "r");
    if (!file) {
        perror("Failed to open input.txt");
        return 1;
    }

    int left[MAX_LINES], right[MAX_LINES];
    int count = 0;

    // Read input
    while (fscanf(file, "%d %d", &left[count], &right[count]) == 2) {
        count++;
        if (count >= MAX_LINES) break;
    }
    fclose(file);

    // Part 1: Sort both lists and calculate total distance
    int left_sorted[MAX_LINES], right_sorted[MAX_LINES];
    memcpy(left_sorted, left, count * sizeof(int));
    memcpy(right_sorted, right, count * sizeof(int));

    qsort(left_sorted, count, sizeof(int), compare_ints);
    qsort(right_sorted, count, sizeof(int), compare_ints);

    long long part1_sum = 0;
    for (int i = 0; i < count; i++) {
        int diff = left_sorted[i] - right_sorted[i];
        part1_sum += diff > 0 ? diff : -diff;
    }

    // Part 2: Calculate similarity score using hash table
    // Build frequency map of right list
    HashEntry *freq_map = calloc(HASH_SIZE, sizeof(HashEntry));
    if (!freq_map) {
        perror("Failed to allocate hash table");
        return 1;
    }

    for (int i = 0; i < count; i++) {
        hash_insert(freq_map, right[i]);
    }

    // Calculate similarity score
    long long part2_sum = 0;
    for (int i = 0; i < count; i++) {
        int num = left[i];
        int occurrences = hash_get(freq_map, num);
        part2_sum += (long long)num * occurrences;
    }

    free(freq_map);

    printf("Part 1: %lld\n", part1_sum);
    printf("Part 2: %lld\n", part2_sum);

    return 0;
}

File: 2024/day01/zig/solution.zig
const std = @import("std");

fn compareInts(_: void, a: i32, b: i32) bool {
    return a < b;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const file_size = (try file.stat()).size;
    const buffer = try allocator.alloc(u8, file_size);
    defer allocator.free(buffer);

    _ = try file.readAll(buffer);

    var left_list: std.ArrayList(i32) = .{};
    defer left_list.deinit(allocator);
    var right_list: std.ArrayList(i32) = .{};
    defer right_list.deinit(allocator);

    // Parse input
    var lines = std.mem.splitScalar(u8, buffer, '\n');
    while (lines.next()) |line| {
        if (line.len == 0) continue;

        var parts = std.mem.splitScalar(u8, line, ' ');
        var nums: std.ArrayList(i32) = .{};
        defer nums.deinit(allocator);

        while (parts.next()) |part| {
            if (part.len == 0) continue;
            const num = try std.fmt.parseInt(i32, part, 10);
            try nums.append(allocator, num);
        }

        if (nums.items.len == 2) {
            try left_list.append(allocator, nums.items[0]);
            try right_list.append(allocator, nums.items[1]);
        }
    }

    // Part 1: Sort both lists and calculate total distance
    const left_sorted = try allocator.dupe(i32, left_list.items);
    defer allocator.free(left_sorted);
    const right_sorted = try allocator.dupe(i32, right_list.items);
    defer allocator.free(right_sorted);

    std.mem.sort(i32, left_sorted, {}, compareInts);
    std.mem.sort(i32, right_sorted, {}, compareInts);

    var part1_sum: i64 = 0;
    for (left_sorted, right_sorted) |l, r| {
        const diff = l - r;
        part1_sum += if (diff > 0) diff else -diff;
    }

    // Part 2: Calculate similarity score using hash map
    var right_counts = std.AutoHashMap(i32, i32).init(allocator);
    defer right_counts.deinit();

    // Build frequency map of right list
    for (right_list.items) |num| {
        const count = right_counts.get(num) orelse 0;
        try right_counts.put(num, count + 1);
    }

    // Calculate similarity score
    var part2_sum: i64 = 0;
    for (left_list.items) |num| {
        const count = right_counts.get(num) orelse 0;
        part2_sum += @as(i64, num) * @as(i64, count);
    }

    std.debug.print("Part 1: {d}\n", .{part1_sum});
    std.debug.print("Part 2: {d}\n", .{part2_sum});
}
